File: python/0056.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals = sorted(intervals, key=lambda intv: intv[0])
        out = [intervals[0]]
        for intv in intervals[1:]:
            if out[-1][1] >= intv[0]:
                out[-1][1] = max(out[-1][1], intv[1])
            else:
                out.append(intv)
        return out

print(Solution().merge([[1, 3], [2, 6], [8, 10], [15, 18]]))
print(Solution().merge([[1, 4], [4, 5]]))

File: cpp/0056.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> res = {intervals[0]};
        for (int i = 1; i < intervals.size(); i++) {
            if (res.back()[1] < intervals[i][0]) {
                res.push_back(intervals[i]);
            } else {
                res.back()[1] = max(res.back()[1], intervals[i][1]);
            }
        }
        return res;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> intervals = {
        {8, 10},
        {2, 6},
        {1, 3},
        {15, 18}
    };
    // vector<vector<int>> intervals = {
    //     {4, 5},
    //     {1, 4}
    // };
    for (auto intv: sol.merge(intervals)) {
        cout << intv[0] << ", " << intv[1] << endl;
    }
    return 0;
}
